numaflow/src/mapstream.rs

```diff
@@ -0,0 +1,887 @@
+    ///
+    /// struct ExampleMapStreamer;
+    ///
+    /// #[async_trait]
+    /// impl MapStreamer for ExampleMapStreamer {
+    ///     async fn map_stream(&self, input: MapStreamRequest, tx: Sender<Message>) {
+    ///         let payload_str = String::from_utf8(input.value).unwrap_or_default();
+    ///         let splits: Vec<&str> = payload_str.split(',').collect();
+    ///
+    ///         for split in splits {
+    ///             let message = Message::new(split.as_bytes().to_vec())
+    ///                 .keys(input.keys.clone())
+    ///                 .tags(vec![]);
+    ///             if tx.send(message).await.is_err() {
+    ///                 break;
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    async fn map_stream(&self, input: MapStreamRequest, tx: Sender<Message>);
+}
+
+/// Message is the response struct from the [`MapStreamer::map_stream`] .
+#[derive(Debug, PartialEq)]
+pub struct Message {
+    /// Keys are a collection of strings which will be passed on to the next vertex as is. It can
+    /// be an empty collection.
+    pub keys: Option<Vec<String>>,
+    /// Value is the value passed to the next vertex.
+    pub value: Vec<u8>,
+    /// Tags are used for [conditional forwarding](https://numaflow.numaproj.io/user-guide/reference/conditional-forwarding/).
+    pub tags: Option<Vec<String>>,
+}
+
+/// Represents a message that can be modified and forwarded.
+impl Message {
+    /// Creates a new message with the specified value.
+    ///
+    /// This constructor initializes the message with no keys, tags, or specific event time.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - A vector of bytes representing the message's payload.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use numaflow::map::Message;
+    /// let message = Message::new(vec![1, 2, 3, 4]);
+    /// ```
+    pub fn new(value: Vec<u8>) -> Self {
+        Self {
+            value,
+            keys: None,
+            tags: None,
+        }
+    }
+    /// Marks the message to be dropped by creating a new `Message` with an empty value and a special "DROP" tag.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use numaflow::map::Message;
+    /// let dropped_message = Message::message_to_drop();
+    /// ```
+    pub fn message_to_drop() -> Message {
+        Message {
+            keys: None,
+            value: vec![],
+            tags: Some(vec![DROP.to_string()]),
+        }
+    }
+
+    /// Sets or replaces the keys associated with this message.
+    ///
+    /// # Arguments
+    ///
+    /// * `keys` - A vector of strings representing the keys.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    ///  use numaflow::map::Message;
+    /// let message = Message::new(vec![1, 2, 3]).keys(vec!["key1".to_string(), "key2".to_string()]);
+    /// ```
+    pub fn keys(mut self, keys: Vec<String>) -> Self {
```

Comment:
use `with_` prefix for methods that set the member fields and return `Self`, like `with_keys()`, `with_tags()`